subsys/net/lib/lwm2m/lwm2m_engine.c

```diff
@@ -0,0 +1,2516 @@
+				    r);
+		}
+
+		r = do_read_op(obj, &context);
+		break;
+
+	case LWM2M_OP_DISCOVER:
+		r = do_discover_op(&context);
+		break;
+
+	case LWM2M_OP_WRITE:
+		r = do_write_op(obj, &context, format);
+		break;
+
+	case LWM2M_OP_WRITE_ATTR:
+		r = lwm2m_write_attr_handler(obj, &context);
+		break;
+
+	case LWM2M_OP_EXECUTE:
+		r = lwm2m_exec_handler(obj, &context);
+		break;
+
+	case LWM2M_OP_DELETE:
+		r = lwm2m_delete_handler(obj, &context);
+		break;
+
+	default:
+		SYS_LOG_ERR("Unknown operation: %u", context.operation);
+		return -EINVAL;
+	}
+
+	if (r == 0) {
+		/* TODO: Handle blockwise 1 */
+
+		if (out.outlen > 0) {
+			SYS_LOG_DBG("replying with %u bytes", out.outlen);
+			zoap_packet_set_used(out.out_zpkt, out.outlen);
+		} else {
+			SYS_LOG_DBG("no data in reply");
+		}
+	} else {
+		if (r == -ENOENT) {
+			zoap_header_set_code(out.out_zpkt,
+					     ZOAP_RESPONSE_CODE_NOT_FOUND);
+			r = 0;
+		} else if (r == -EPERM) {
+			zoap_header_set_code(out.out_zpkt,
+					     ZOAP_RESPONSE_CODE_NOT_ALLOWED);
+			r = 0;
+		} else {
+			/* Failed to handle the request */
+			zoap_header_set_code(out.out_zpkt,
+					     ZOAP_RESPONSE_CODE_INTERNAL_ERROR);
+			r = 0;
+		}
+	}
+
+	return r;
+}
+
+void lwm2m_udp_receive(struct net_context *ctx, struct net_pkt *pkt,
+		       struct zoap_pending *zpendings, int num_zpendings,
+		       struct zoap_reply *zreplies, int num_zreplies,
+		       int (*udp_request_handler)(struct zoap_packet *,
+						  struct zoap_packet *,
+						  struct sockaddr *))
+{
+	struct net_udp_hdr hdr, *udp_hdr;
+	struct zoap_pending *pending;
+	struct zoap_reply *reply;
+	struct zoap_packet response;
+	struct sockaddr from_addr;
+	struct zoap_packet response2;
+	struct net_pkt *pkt2;
+	int header_len, r;
+	const u8_t *token;
+	u8_t tkl;
+
+	udp_hdr = net_udp_get_hdr(pkt, &hdr);
+	if (!udp_hdr) {
+		SYS_LOG_ERR("Invalid UDP data");
+		return;
+	}
+
+	/* Save the from address */
+#if defined(CONFIG_NET_IPV6)
+	if (net_pkt_family(pkt) == AF_INET6) {
+		net_ipaddr_copy(&net_sin6(&from_addr)->sin6_addr,
+				&NET_IPV6_HDR(pkt)->src);
+		net_sin6(&from_addr)->sin6_port = udp_hdr->src_port;
+		net_sin6(&from_addr)->sin6_family = AF_INET6;
+	}
+#endif
+
+#if defined(CONFIG_NET_IPV4)
+	if (net_pkt_family(pkt) == AF_INET) {
+		net_ipaddr_copy(&net_sin(&from_addr)->sin_addr,
+				&NET_IPV4_HDR(pkt)->src);
+		net_sin(&from_addr)->sin_port = udp_hdr->src_port;
+		net_sin(&from_addr)->sin_family = AF_INET;
+	}
+#endif
+
+	/*
+	 * zoap expects that buffer->data starts at the
+	 * beginning of the CoAP header
+	 */
+	header_len = net_pkt_appdata(pkt) - pkt->frags->data;
+	net_buf_pull(pkt->frags, header_len);
+
+	r = zoap_packet_parse(&response, pkt);
+	if (r < 0) {
+		SYS_LOG_ERR("Invalid data received (err:%d)", r);
+		goto cleanup;
+	}
+
+	token = zoap_header_get_token(&response, &tkl);
+	pending = zoap_pending_received(&response, zpendings, num_zpendings);
+	if (pending) {
+		/* If necessary cancel retransmissions */
```

Comment:
Is this to be implemented or just missing atm? Please add "TODO"  here if this needs to be implemented later.